addons/account/models/account_move.py

```diff
@@ -1581,41 +1645,21 @@ def _check_journal_move_type(self):
 
     @api.constrains('ref', 'move_type', 'partner_id', 'journal_id', 'invoice_date', 'state')
     def _check_duplicate_supplier_reference(self):
-        moves = self.filtered(lambda move: move.state == 'posted' and move.is_purchase_document() and move.ref)
-        if not moves:
-            return
-
-        self.env["account.move"].flush_model([
-            "ref", "move_type", "invoice_date", "journal_id",
-            "company_id", "partner_id", "commercial_partner_id",
-        ])
-        self.env["account.journal"].flush_model(["company_id"])
-        self.env["res.partner"].flush_model(["commercial_partner_id"])
-
-        # /!\ Computed stored fields are not yet inside the database.
-        self._cr.execute('''
-            SELECT move2.id
-            FROM account_move move
-            JOIN account_journal journal ON journal.id = move.journal_id
-            JOIN res_partner partner ON partner.id = move.partner_id
-            INNER JOIN account_move move2 ON
-                move2.ref = move.ref
-                AND move2.company_id = journal.company_id
-                AND move2.commercial_partner_id = partner.commercial_partner_id
-                AND move2.move_type = move.move_type
-                AND (move.invoice_date is NULL OR move2.invoice_date = move.invoice_date)
-                AND move2.id != move.id
-            WHERE move.id IN %s
-        ''', [tuple(moves.ids)])
-        duplicated_moves = self.browse([r[0] for r in self._cr.fetchall()])
-        if duplicated_moves:
-            raise ValidationError(_('Duplicated vendor reference detected. You probably encoded twice the same vendor bill/credit note:\n%s') % "\n".join(
-                duplicated_moves.mapped(lambda m: "%(partner)s - %(ref)s - %(date)s" % {
-                    'ref': m.ref,
-                    'partner': m.partner_id.display_name,
-                    'date': format_date(self.env, m.invoice_date),
-                })
+        """ Assert the move which is about to be posted isn't a duplicated move from another posted entry"""
+        move_to_duplicate_moves = self.filtered(lambda m: m.state == 'posted')._fetch_duplicate_supplier_reference(only_posted=True)
+        if any(m for m in move_to_duplicate_moves.values()):
+            duplicated_moves = self.env['account.move'].browse(set(
+                move.id
+                for moves in move_to_duplicate_moves.values()
+                for move in moves
             ))
+            raise ValidationError(
+                _('Duplicated vendor reference detected. You probably encoded twice the same vendor bill/credit note:\n%s')
+                % "\n".join(
+                    duplicated_moves.sorted(lambda m: (m.commercial_partner_id.id, m.invoice_date, ))
+                        .mapped(lambda m: f"{m.partner_id.display_name} - {m.ref} - {format_date(self.env, m.invoice_date)} - {format_amount(self.env, m.amount_total, m.currency_id)}")
+                )
+            )
```

Comment:
Redirect warning with a group by ref (maybe followed by a group by date?; I think that it is possible with numbers) in the context?

This means we don't need all the logic in this function anymore